### PR TITLE
Feature/erutaef

### PR DESCRIPTION
**Pull request by scoreyou**

### CODE REVIEW

File: LeadKitAdditions/CommonCrypto/CommonCrypto.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
+//
```
Comment (NikAshanin) [inline LeadKitAdditions/CommonCrypto/CommonCrypto.h]: copyright as in leadkit

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+//  THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+public struct PassCodeConfiguration {
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: open?

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public var passCodeCharactersNumber: UInt = 4
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: open?

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    public var maxAttemptsLoginNumber: UInt = 5
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: open?

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        guard passCodeCharactersNumber > 0 else {
+            assertionFailure("passCodeCharactersNumber must be greater then 0")
+            return nil
+        }
+        self.passCodeCharactersNumber = passCodeCharactersNumber
+    }
+
+    public static var defaultConfiguration: PassCodeConfiguration {
+        let passCodeConfiguration = PassCodeConfiguration()
+        return passCodeConfiguration
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: return PassCodeConfiguration() ?

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeError.swift [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+public enum PassCodeError: Error {
+    case codesNotMatch
+    case wrongCode
+    case tooMuchAttempts
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeError.swift]: tooManyAttempts?

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeHolder.swift [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        return self as? PassCodeHolderCreate
+    }
+
+    public var passCodeHolderEnter: PassCodeHolderEnter? {
+        return self as? PassCodeHolderEnter
+    }
+
+}
+
+public class PassCodeHolderCreate: PassCodeHolderProtocol {
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeHolder.swift]: open?

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift (hunk 2 above)
Comment (scoreyou) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: Only classes and overridable class members can be declared 'open'

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift (hunk 4 above)
Comment (scoreyou) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: Only classes and overridable class members can be declared 'open'

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift (hunk 3 above)
Comment (scoreyou) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeConfiguration.swift]: Only classes and overridable class members can be declared 'open'

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeHolder.swift (hunk 7 above)
Comment (scoreyou) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/Model/PassCodeHolder.swift]: Nope, this implementation is final

File: LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/ViewModel/BasePassCodeViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        validationResultHolder.value = nil
+        passCodeControllerStateHolder.value = .enter
+        attemptsNumber = 0
+        passCodeHolder.reset()
+    }
+
+    // MARK: - HAVE TO OVERRIDE
+
+    open func isEnteredPassCodeValid(_ passCode: String) -> Bool {
+        assertionFailure("Don't use it directly. Override it!")
```
Comment (NikAshanin) [inline LeadKitAdditions/LeadKitAdditions/Controllers/PassCode/ViewModel/BasePassCodeViewModel.swift]: better to write what method should be overrided
do it everywhere

**Pull request merged**